libsvmdata/datasets.py

```diff
@@ -349,30 +380,44 @@ def _get_X_y(dataset, multilabel, replace=False, verbose=False):
         n_features_total = N_FEATURES[dataset]
         if verbose:
             print("Loading svmlight file...")
-        with open(tmp_path, 'rb') as f:
-            X, y = load_svmlight_file(
-                f, n_features=n_features_total, multilabel=multilabel)
-
-        tmp_path.unlink()
-        # if X's density is more than 0.5, store it in dense format:
-        if len(X.data) >= 0.5 * X.shape[0] * X.shape[1]:
-            X = X.toarray(order='F')
-            np.save(X_path, X)
-        else:
-            X = sparse.csc_matrix(X)
-            X.sort_indices()
-            sparse.save_npz(X_path, X)
-
-        if multilabel:
-            indices = np.array([lab for labels in y for lab in labels])
-            indptr = np.cumsum([0] + [len(labels) for labels in y])
-            data = np.ones_like(indices)
-            Y = sparse.csr_matrix((data, indices, indptr))
-            sparse.save_npz(y_path, Y)
-            return X, Y
 
-        else:
-            np.save(y_path, y)
+        # Loop over multiple files to cope with cases where the dataset is in
+        # an archive. Store the result for all datasets in the archive and
+        # keep the result to return in X_y_.
+        X_y_ = None
+        for path, d in path_mapping.items():
+            with open(path, 'rb') as f:
+                X, y = load_svmlight_file(
+                    f, n_features=n_features_total, multilabel=multilabel
+                )
+
+            if d == dataset:
```

Comment:
shouldn't this go L417 ? So as to return a dense X when X has more than 50% non zero elements, and also to handle the multilabel case ?